File: prototype_python/player.py
"""Player data model with fields that map cleanly to a future C struct."""

from .constants import (
    DEFAULT_STARTING_LIFE,
    MAX_COMMANDERS_PER_PLAYER,
    MAX_CUSTOM_COUNTERS,
    MAX_NAME_LENGTH,
)


class Player:
    """Mutable state owned by one player."""

    def __init__(
        self,
        player_id: int,
        name: str,
        starting_life: int = DEFAULT_STARTING_LIFE,
    ) -> None:
        if player_id < 0:
            raise ValueError("player_id cannot be negative")
        if not name:
            raise ValueError("player name cannot be empty")
        if len(name) > MAX_NAME_LENGTH:
            raise ValueError(f"player name cannot exceed {MAX_NAME_LENGTH} characters")

        self.player_id = player_id
        self.name = name
        self.life = starting_life
        self.poison = 0
        self.rad = 0
        self.energy = 0
        self.experience = 0
        self.ticket = 0
        self.eliminated = False
        self.commander_tax = [0] * MAX_COMMANDERS_PER_PLAYER
        self.commander_cast_count = [0] * MAX_COMMANDERS_PER_PLAYER
        self.custom_counters = [0] * MAX_CUSTOM_COUNTERS

        # Unique statuses are mirrored here for convenient player display. GameState
        # remains authoritative and ensures that only one holder exists.
        self.is_monarch = False
        self.has_initiative = False
        self.has_citys_blessing = False

    def reset(self, starting_life: int) -> None:
        """Reset gameplay values while keeping the player's ID and name."""
        self.life = starting_life
        self.poison = 0
        self.rad = 0
        self.energy = 0
        self.experience = 0
        self.ticket = 0
        self.eliminated = False
        self.commander_tax = [0] * MAX_COMMANDERS_PER_PLAYER
        self.commander_cast_count = [0] * MAX_COMMANDERS_PER_PLAYER
        self.custom_counters = [0] * MAX_CUSTOM_COUNTERS
        self.is_monarch = False
        self.has_initiative = False
        self.has_citys_blessing = False

File: gbc_src/player.c
#include "player.h"

static void copy_short_name(char *destination, const char *source) {
    uint8_t index = 0u;

    while ((index < MAX_NAME_LENGTH) && (source[index] != '\0')) {
        destination[index] = source[index];
        index++;
    }
    destination[index] = '\0';
}

void player_reset(Player *player, int16_t starting_life) {
    uint8_t index;

    player->life = starting_life;
    player->poison = 0u;
    player->rad = 0u;
    player->energy = 0u;
    player->experience = 0u;
    player->ticket = 0u;
    player->commander_id = NO_COMMANDER_ID;
    player->eliminated = 0u;
    player->is_monarch = 0u;
    player->has_initiative = 0u;
    player->has_citys_blessing = 0u;

    for (index = 0u; index < MAX_COMMANDERS_PER_PLAYER; index++) {
        player->commander_tax[index] = 0u;
        player->commander_cast_count[index] = 0u;
    }
    for (index = 0u; index < MAX_CUSTOM_COUNTERS; index++) {
        player->custom_counters[index] = 0u;
    }
}

void player_init(
    Player *player,
    uint8_t player_id,
    const char *name,
    int16_t starting_life
) {
    player->player_id = player_id;
    copy_short_name(player->name, name);
    player_reset(player, starting_life);
}
